### DISCUSSION TITLE
[Suggestion] Nix/NixOS containers

### DISCUSSION THREAD

**Discussion by slopsec** [89luca89/distrobox#1886] [Ideas]
**Is your feature request related to a problem? Please describe.**

Nix and NixOS are entirely different beasts and there are circumstances where it would be beneficial to have NixOS on your system in a containerized format. Nix when installed tends to embed itself into your system in a manner that is difficult to remove and does not leverage the full strength of Nix and it's declarative format. Some services you may want to run under Nix may require a NixOS module to use which can significantly limit how Nix is deployed. I believe that by providing a NixOS container, you can address the shortcomings of using Nix on a non-NixOS Linux distribution while allowing people to more easily leverage Nix through NixOS modules. 

**Describe the solution you'd like**
Support for running Nix/NixOS container images. 

**Describe alternatives you've considered**
I've attempted to use a Nix/NixOS container image in Docker/Podman software made to make it easier to interface and use containerization, though I've found that no solutions come close to Distrobox.

**Additional context**
Nix isn't supported on RHEL-derivative Distributions that make of SELLinux which significantly limit where it can be deployed, supporting Nix/NixOS containers in Distrobox could help in that respect.

**Comment by dottorblaster** [comment]
I would be interested in seeing this going forward, I'd like to move it to discussions in order to start accepting PRs for this.

**Reply by slopsec** [reply]
I've made some attempts at getting it to work previously but have noticed that the Nix/NixOS docker image is pretty broken. This was just using the image itself with regular docker/podman. I'm not sure whether @89luca89 or anyone else would be interested in inheriting the technical debt that would come from cleaning it up to be usable out of the box.

I was encouraged at that point to just use systemd-nspawn instead in an unofficial Nix community. Interested as I was to see this developed further myself, I do think it is important that people be made aware that it may prove to be troublesome to get working.

@dottorblaster

**Comment by ricardoseriani** [comment]
It would be awesome to have support for Nix/NixOS containers, as they provide many packages!

I don't know about the problems that @slopsec has, because I'm not running the official docker image.

I have an ubuntu box with nix package manager installed, and I use that to install the nix packages that I want.

However, there are some downsides in doing this:
- Some packages, like wezterm, require [nixGL](https://github.com/nix-community/nixGL) to run, so when I export the apps/binaries, I have to manually edit to add this. (I have already opened a discussion about this in https://github.com/89luca89/distrobox/discussions/1827).
- The update/upgrade of packages is manually (I can't use distrobox upgrade).

Of that difficulties, the one that bothers me most is the one with nixGL.

I knew that this discussion isn't about this, but if you're not using the official NixOS image, I think that will need some any way to prepend a command on the exported apps/binaries.

**Reply by slopsec** [reply]
> I don't know about the problems that @slopsec has, because I'm not running the official docker image.

It was mostly trying to get the Nix/NixOS image to behave and work properly. Out of the box, you can't really use the image as you would on NixOS, which I imagined was the whole point of shipping the image as Nix is available elsewhere as a package manager. The difference between Nix as a package manager and NixOS as a distribution is significant enough in my opinion to warrant having a container image made available for it, though. Once you've daily driven NixOS, going back to just Nix feels like you've lost a lot of that functionality you previously had with modules.
